### PR TITLE
Add rudimental syscall capget/capset

### PR DESCRIPTION
**Pull request by StanPlatinum**
Add the capget/capset syscalls & Linux capability support (for prctl later).

Also, a test application has been added to `regression/apps/capability`.

### CODE REVIEW

**Review by tatetian — COMMENTED** [review]
Thank you for submitting the implementation of Linux capabilities.

My biggest concern is that the implementation proposed by this PR appears to be too rudimental. I call it "rudimental" because it lacks of the essence of the Linux capabilities mechanism. Adding the two system calls is only a good first step, but not enough. The essence of the Linux capabilities, in my opinion, is enforcing its security invariants. They should have been have been done by this PR.

To describe the invariants, I follow the notations adopted by the man page.

* `P(effective)`, `P(permitted)`, and `P(inheritable)` denote the effective, permitted, and inheritable capability sets of a process, respectively.
* `P'(effective)`, `P'(permitted)`, and `P'(inheritable)` denote the effective, permitted, and inheritable capability sets of a forked or exec'ed process, respectively.
* `F(effective)`, `F(permitted)`, and `F(inheritable)` denote the effective, permitted, and inheritable capability sets of a file exec'ed by a process, respectively. 

The essence of the Linux capabilities system include but is not limited to the following security invariants categorized by their usage. None of these invariants are enforced by this PR.

## When determining whether a privilege operation is allowed or not

**Invariant.** A thread is only allowed to perform privileged operations granted by the capabilities within its `P(effective)` set.

This invariant must be enforced by changing the specific subsystems or modules that actually conduct the privilege operations. This is too much work and thus can be left out of this PR.

Note that Linux capabilities are per-thread, not per-process. This is a point that is NOT implemented faithfully by this PR.

## When passing on capabilities from a parent process to a child on `fork` or `clone`

**Invariant:** `P(effective) = P'(effective)`

**Invariant:** `P(permitted) = P'(permitted)`

**Invariant:** `P(inheritable) = P'(inheritable)`

These invariants are trivial. They basically say that capabilities are passed intact to a child.

## When adding or dropping the capabilities of a process/thread

* **Invariant:** `P(effective) <= P(permitted)`
* **Invariant:** `P(inheritable) <= P(permitted)`
* **Invariant:**  `P(permitted) >= P'(permitted)`

Here, `P'(permitted)` means the new value of the permitted capability set.

## When execing a file

* **Invariant:** `P'(permitted) = (P(inheritable) & F(inheritable)) | (F(permitted) & cap_bset)`
* **Invariant:** `P'(effective) = F(effective) ? P'(permitted) : 0`
* **Invariant:** `P'(inheritable) = P(inheritable)`

`cap_bset` denotes the process bounding capability set.

**Comment by StanPlatinum** [comment]
> Thank you for submitting the implementation of Linux capabilities.
> 
> My biggest concern is that the implementation proposed by this PR appears to be too rudimental. I call it "rudimental" because it lacks of the essence of the Linux capabilities mechanism. Adding the two system calls is only a good first step, but not enough. The essence of the Linux capabilities, in my opinion, is enforcing its security invariants. They should have been have been done by this PR.
> 
> To describe the invariants, I follow the notations adopted by the man page.
> 
> * `P(effective)`, `P(permitted)`, and `P(inheritable)` denote the effective, permitted, and inheritable capability sets of a process, respectively.
> * `P'(effective)`, `P'(permitted)`, and `P'(inheritable)` denote the effective, permitted, and inheritable capability sets of a forked or exec'ed process, respectively.
> * `F(effective)`, `F(permitted)`, and `F(inheritable)` denote the effective, permitted, and inheritable capability sets of a file exec'ed by a process, respectively.
> 
> The essence of the Linux capabilities system include but is not limited to the following security invariants categorized by their usage. None of these invariants are enforced by this PR.
> 
> ## When determining whether a privilege operation is allowed or not
> * **Invariant.** A thread is only allowed to perform privileged operations granted by the capabilities within its `P(effective)` set.
> 
> This invariant must be enforced by changing the specific subsystems or modules that actually conduct the privilege operations. This is too much work and thus can be left out of this PR.
> 
> Note that Linux capabilities are per-thread, not per-process. This is a point that is NOT implemented faithfully by this PR.
> 
> ## When passing on capabilities from a parent process to a child on `fork` or `clone`
> * **Invariant:** `P'(effective) = P(effective)`
> * **Invariant:** `P'(permitted) = P(permitted)`
> * **Invariant:** `P'(inheritable) = P(inheritable)`
> 
> These invariants are trivial. They basically say that capabilities are passed intact to a child.
> 
> ## When adding or dropping the capabilities of a process/thread
> The `capset` system call can be used to adjust the capability sets of a process/thread. The fundamental rule is that the permitted capability set is the superset of all current and future capability sets (until the thread execes a file).
> 
> * **Invariant:** `P(effective) <= P(permitted)`
> * **Invariant:** `P(inheritable) <= P(permitted)`
> * **Invariant:**  `P'(permitted) <= P(permitted)`
> 
> Here, `P'(permitted)` means the new value of the permitted capability set.
> 
> ## When execing a file
> * **Invariant:** `P'(permitted) = (P(inheritable) & F(inheritable)) | (F(permitted) & cap_bset)`
> * **Invariant:** `P'(effective) = F(effective) ? P'(permitted) : 0`
> * **Invariant:** `P'(inheritable) = P(inheritable)`
> 
> `cap_bset` denotes the process bounding capability set.
> 
> ## More
> There are more invariants to enforce. I only summarize some of them as I read through _Chapter 39 Capabilities, The Linux Programming Interface_. I expect you to dive into the spec more deeply than I do to come up with a faithful and correct implementation.

Thanks! Get it. I will add a more correct implementation in the later PR(s).

File: kernel/aster-nix/src/process/credentials/static_cap.rs [hunk 1]
```diff
@@ -247,4 +248,54 @@ impl<R: TRights> Credentials<R> {
...
+
+    /// Sets the Linux caps that children can inherit.
+    ///
+    /// This method requires the `Write` right.
+    #[require(R > Write)]
+    pub fn set_inheritablecap(&self, inheritablecap: Capability) {
+        self.0.set_inheritablecap(inheritablecap);
+    }
+
+    /// Sets the Linux caps that caps are permitted.
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/static_cap.rs:286]: Done.

File: kernel/aster-nix/src/syscall/capset.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            linuxcapability::{Capability, CAP_VALID_MASK},
+        },
+        credentials_mut,
+    },
+    util::read_val_from_user,
+};
+
+/// Convert the high 32 bits to a u64 and shift them left by 32 bits,
+/// then combine them with the low 32 bits, and apply a mask.
+fn mk_kernel_cap(low: u32, high: u32) -> u64 {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/syscall/capset.rs:18]: Revised.

File: kernel/aster-nix/src/syscall/capset.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+pub fn sys_capset(cap_user_header_addr: Vaddr, cap_user_data_addr: Vaddr) -> Result<SyscallReturn> {
+    let cap_user_header: cap_user_header_t =
+        read_val_from_user::<cap_user_header_t>(cap_user_header_addr)?;
+
+    // Validate the `header`.
+    if cap_user_header.version != LINUX_CAPABILITY_VERSION_3 {
+        // Return an error if the version is not supported.
+        // For simplicity, we assume that the header version is CAPABILITY_VERSION_3.
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/syscall/capset.rs:29]: Done.

File: kernel/aster-nix/src/process/credentials/linuxcapability.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+impl Clone for AtomicCapability {
+    fn clone(&self) -> Self {
+        Self(AtomicU64::new(self.0.load(Ordering::Relaxed)))
+    }
+}
+
+// Structure that represents the different capability sets used by the Linux capability model.
+pub struct CapabilitySets {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/linuxcapability.rs:130]: Removed.

File: kernel/aster-nix/src/process/credentials/static_cap.rs [hunk 5]
```diff
@@ -247,4 +248,54 @@ impl<R: TRights> Credentials<R> {
...
         self.0.groups_mut()
     }
+
+    // *********** Linux Capability methods **********
+
+    /// Gets the Linux caps that children can inherit.
+    ///
+    /// This method requies the `Read` right.
+    #[require(R > Read)]
+    pub fn inheritablecap(&self) -> Capability {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/static_cap.rs:258]: Done.

File: regression/apps/capability/check_cap.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		printf("Process has CAP_NET_RAW capability.\n");
+	} else {
+		printf("Process does NOT have CAP_NET_RAW capability.\n");
+	}
+
+	// check CAP_NET_ADMIN
+	if (check_capability(1 << CAP_NET_ADMIN, target_pid)) {
+		printf("Process has CAP_NET_ADMIN capability.\n");
+	} else {
+		printf("Process does NOT have CAP_NET_ADMIN capability.\n");
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply regression/apps/capability/check_cap.c:75]: Revised.

File: kernel/aster-nix/src/process/credentials/linuxcapability.rs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+use core::sync::atomic::{AtomicU64, Ordering};
+
+use crate::prelude::*;
+
+pub type CapMask = u64;
+
+// Capabilities defined by Linux. Taken from the kernel's include/uapi/linux/capability.h.
+// See capabilities(7) or that file for more detailed capability descriptions.
+#[repr(u8)]
+pub enum CapabilityPrivilegeList {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/linuxcapability.rs]: Revised.

File: kernel/aster-nix/src/process/credentials/linuxcapability.rs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    pub const fn new_root() -> Self {
+        // Give CAP_SYS_ADMIN for the root.
+        Self(1u64 << CapabilityPrivilegeList::SysAdmin as u64)
+    }
+
+    // Provided for prctl(PR_SET_KEEPCAPS), prctl(PR_CAP_AMBIENT_RAISE)
+    // Sets (raises) a particular capability bit in a capability set, identified by the `flag` index.
+    pub fn raise(cap_set: &mut CapMask, flag: u64) {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/linuxcapability.rs]: Revised.

File: kernel/aster-nix/src/process/credentials/credentials_.rs [hunk 9]
```diff
@@ -25,13 +28,27 @@ pub(super) struct Credentials_ {
...
     fsgid: AtomicGid,
 
-    // A set of additional groups to which a process belongs.
+    /// A set of additional groups to which a process belongs.
     supplementary_gids: RwLock<BTreeSet<Gid>>,
+
+    /// The Linux capabilities.
+    /// This is not the capability (in static_cap.rs) enforced on rust objects.
+
+    /// Capability that children can inherit
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/credentials_.rs]: Done.

File: kernel/aster-nix/src/process/credentials/static_cap.rs [hunk 10]
```diff
@@ -247,4 +248,54 @@ impl<R: TRights> Credentials<R> {
...
         self.0.groups_mut()
     }
+
+    // *********** Linux Capability methods **********
+
+    /// Gets the capabilities that children can inherit.
+    ///
+    /// This method requies the `Read` right.
+    #[require(R > Read)]
+    pub fn inheritable_cap(&self) -> CapSet {
```
[the remark this answers is not in the document]
Comment (StanPlatinum) [reply kernel/aster-nix/src/process/credentials/static_cap.rs]: Revised.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**